infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java

```diff
@@ -0,0 +1,211 @@
+import io.fabric8.kubernetes.api.model.Volume;
+import io.fabric8.kubernetes.api.model.VolumeBuilder;
+import io.fabric8.kubernetes.api.model.VolumeMountBuilder;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.Map.Entry;
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment.PodData;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment.PodRole;
+import org.eclipse.che.workspace.infrastructure.kubernetes.namespace.KubernetesNamespace;
+
+/**
+ * Finds secrets with specific labels in namespace, and mount their values as file or environment
+ * variable into all (or specified by "org.eclipse.che/target-container" annotation) workspace
+ * containers. Secrets with annotation "org.eclipse.che/mount-as=env" are mount as env variables,
+ * env name is read from "org.eclipse.che/env-name" annotation. Secrets which don't have
+ * "org.eclipse.che/mount-as=env" or having "org.eclipse.che/mount-as=file" are mounted as file in
+ * the folder specified by "org.eclipse.che/mount-path" annotation. Refer to che-docs for concrete
+ * examples.
+ */
+@Beta
+@Singleton
+public class SecretAsContainerResourceProvisioner<E extends KubernetesEnvironment> {
+
+  private static final String ANNOTATION_PREFIX = "che.eclipse.org";
+  static final String ANNOTATION_MOUNT_AS = ANNOTATION_PREFIX + "/" + "mount-as";
+  static final String ANNOTATION_TARGET_CONTAINER = ANNOTATION_PREFIX + "/" + "target-container";
+  static final String ANNOTATION_ENV_NAME = ANNOTATION_PREFIX + "/" + "env-name";
+  static final String ANNOTATION_ENV_NAME_TEMPLATE = ANNOTATION_PREFIX + "/%s_" + "env-name";
+  static final String ANNOTATION_MOUNT_PATH = ANNOTATION_PREFIX + "/" + "mount-path";
+
+  private final Map<String, String> secretLabels;
+
+  @Inject
+  public SecretAsContainerResourceProvisioner(
+      @Named("che.workspace.provision.secret.labels") String[] labels) {
+    this.secretLabels =
+        Arrays.stream(labels)
+            .map(item -> item.split("=", 2))
+            .collect(toMap(p -> p[0], p -> p.length == 1 ? "" : p[1]));
+  }
+
+  public void provision(E env, KubernetesNamespace namespace) throws InfrastructureException {
+    LabelSelector selector = new LabelSelectorBuilder().withMatchLabels(secretLabels).build();
+    for (Secret secret : namespace.secrets().get(selector)) {
+      String targetContainerName =
+          secret.getMetadata().getAnnotations().get(ANNOTATION_TARGET_CONTAINER);
+      if ("env".equalsIgnoreCase(secret.getMetadata().getAnnotations().get(ANNOTATION_MOUNT_AS))) {
+        mountAsEnv(env, secret, targetContainerName);
+      } else {
```

Comment:
are we ok with `.../mount-as: boo` being interpreted as "mount as file"?

IMHO, this should warn if the annotation is anything else but the supported values and only interpret the lack of the annotation as the default value - mount as file.

Reply:
Made it strict validated